### PR TITLE
feat: Change EKS default version to 1.17

### PR DESCRIPTION
**Pull request by sc250024**
# PR o'clock

## Description

Upgrading module to support EKS 1.17. Resolves issue https://github.com/terraform-aws-modules/terraform-aws-eks/issues/947.

### Checklist

- [x] CI tests are passing
- [x] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

### CODE REVIEW

**Comment by dpiddockcmp** [comment]
Pushing out a major version every time AWS releases a new version seems a little silly to me. Can't we just make `cluster_version` a required parameter, no default? One last major version bump for this particular issue. Then we'd only need to update the examples which can be a minor version bump.

@barryib @max-rocket-internet what do you think?

**Comment by sc250024** [comment]
Just rebased against the latest `master` branch. Re-review needed.

**Comment by gdurandvadas** [comment]
One note on this. When using `node_groups`, updating the EKS version to `1.17` doesn't update the node groups AMI version.
Moreover, the AMI version required to work with EKS 1.17 ([1.17.7-20200709](https://docs.aws.amazon.com/eks/latest/userguide/eks-linux-ami-versions.html)) is not allowed to be used.
```
Error: error updating EKS Node Group (main-development:main-development-system-1-keen-glider) version: InvalidParameterException: Requested Nodegroup release version 1.17.7-20200709 is invalid. Allowed release version is 1.16.12-20200710
{
  RespMetadata: {
    StatusCode: 400,
    RequestID: "897202ad-0a43-40d0-a5b8-a2cb72211ee3"
  },
  ClusterName: "main-development",
  Message_: "Requested Nodegroup release version 1.17.7-20200709 is invalid. Allowed release version is 1.16.12-20200710",
  NodegroupName: "main-development-system-1-keen-glider"
}
```

**Comment by sc250024** [comment]
> One note on this. When using `node_groups`, updating the EKS version to `1.17` doesn't update the node groups AMI version.
> Moreover, the AMI version required to work with EKS 1.17 ([1.17.7-20200709](https://docs.aws.amazon.com/eks/latest/userguide/eks-linux-ami-versions.html)) is not allowed to be used.
> 
> ```
> Error: error updating EKS Node Group (main-development:main-development-system-1-keen-glider) version: InvalidParameterException: Requested Nodegroup release version 1.17.7-20200709 is invalid. Allowed release version is 1.16.12-20200710
> {
>   RespMetadata: {
>     StatusCode: 400,
>     RequestID: "897202ad-0a43-40d0-a5b8-a2cb72211ee3"
>   },
>   ClusterName: "main-development",
>   Message_: "Requested Nodegroup release version 1.17.7-20200709 is invalid. Allowed release version is 1.16.12-20200710",
>   NodegroupName: "main-development-system-1-keen-glider"
> }
> ```

I'm not sure how to fix this because the AMI version is retrieved automatically from `local.worker_ami_name_filter`. Did they change their naming schema on the AMIs?

**Comment by sc250024** [comment]
@gdurandvadas It might be related to this, in which case I'd need to bump the provider version: https://github.com/terraform-providers/terraform-provider-aws/issues/13407

**Comment by sc250024** [comment]
> Pushing out a major version every time AWS releases a new version seems a little silly to me. Can't we just make `cluster_version` a required parameter, no default? One last major version bump for this particular issue. Then we'd only need to update the examples which can be a minor version bump.
> 
> @barryib @max-rocket-internet what do you think?

A counterargument on this: look at the comment from @gdurandvadas above. He discovered that when upgrading his cluster to 1.17, he ran into a problem.

Now, the problem wasn't related to this module; instead, it was a problem with the version of the AWS provider he was using, and was fixed in provider version `v2.63.0`. By having regular major version releases, we can address problems like this; i.e. bumping the major version of this module gives an nice opportunity to make other breaking changes, if needed.

**Comment by gdurandvadas** [comment]
@sc250024 thanks for the update, agreed that this is not the module itself.
My `versions.tf` has defined `aws = ">= 2.52.0"` but if I check the _.terraform_ folder, I see the downloaded plugin is `terraform-provider-aws_v2.66.0_x4`
Let me try with the latest AWS provider version (v2.70.0). I'll come back to you.

**Comment by gdurandvadas** [comment]
> 
> The AMI version is retrieved automatically from `local.worker_ami_name_filter`, so it must be something else.
> 
> @gdurandvadas What AWS Terraform provider version are you using?

Bear in mind that I'm using `node_groups` not workers, and I set `ami_release_version=1.17.7-20200709` on the node definition.

**Comment by dpiddockcmp** [comment]
We can, and do, put out major versions when other breaking changes are made. I don't think that AWS releasing a new supported version of kubernetes in EKS should trigger the module to release a major version. This is unnecessary work for users.

Looking through the changelog, I think the following releases were major version bumps just because of the kubernetes version:
- [v3.0.0](https://github.com/terraform-aws-modules/terraform-aws-eks/compare/v2.3.1...v3.0.0) 1.12
- [v10.0.0](https://github.com/terraform-aws-modules/terraform-aws-eks/compare/v9.0.0...v10.0.0) 1.15
- [v12.0.0](https://github.com/terraform-aws-modules/terraform-aws-eks/compare/v11.1.0...v12.0.0) 1.16

OK, that's less than I would have thought. My memory must be biased due to the previous two kubernetes releases. [1.11](https://github.com/terraform-aws-modules/terraform-aws-eks/compare/v1.8.0...v2.0.0), [1.13](https://github.com/terraform-aws-modules/terraform-aws-eks/compare/v4.0.2...v5.0.0) and [1.14](https://github.com/terraform-aws-modules/terraform-aws-eks/compare/v5.1.0...v6.0.0) were mixed in with other big changes. But having a major release pending does seem to delay the potential for a release. Mainly because we only run a single branch.

I still think having the version set as a default is a little silly 😉

**Comment by sc250024** [comment]
> We can, and do, put out major versions when other breaking changes are made. I don't think that AWS releasing a new supported version of kubernetes in EKS should trigger the module to release a major version. This is unnecessary work for users.
> 
> Looking through the changelog, I think the following releases were major version bumps just because of the kubernetes version:
> 
> * [v3.0.0](https://github.com/terraform-aws-modules/terraform-aws-eks/compare/v2.3.1...v3.0.0) 1.12
> * [v10.0.0](https://github.com/terraform-aws-modules/terraform-aws-eks/compare/v9.0.0...v10.0.0) 1.15
> * [v12.0.0](https://github.com/terraform-aws-modules/terraform-aws-eks/compare/v11.1.0...v12.0.0) 1.16
> 
> OK, that's less than I would have thought. My memory must be biased due to the previous two kubernetes releases. [1.11](https://github.com/terraform-aws-modules/terraform-aws-eks/compare/v1.8.0...v2.0.0), [1.13](https://github.com/terraform-aws-modules/terraform-aws-eks/compare/v4.0.2...v5.0.0) and [1.14](https://github.com/terraform-aws-modules/terraform-aws-eks/compare/v5.1.0...v6.0.0) were mixed in with other big changes. But having a major release pending does seem to delay the potential for a release. Mainly because we only run a single branch.
> 
> I still think having the version set as a default is a little silly 😉

@dpiddockcmp In the end, it's your guys' module. I don't mind the version bump personally, but if it's decided to get rid of them, then 🤷‍♂️.

@filipedeo @barryib Is there anything else needed for this to get merged?

File: README.md [hunk 1]
```diff
@@ -15,11 +15,19 @@ Read the [AWS docs on EKS to get connected to the k8s dashboard](https://docs.aw
 * You want these resources to exist within security groups that allow communication and coordination. These can be user provided or created within the module.
 * You've created a Virtual Private Cloud (VPC) and subnets where you intend to put the EKS resources. The VPC satisfies [EKS requirements](https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html).
 
-## Important note
```
Comment (barryib) [inline README.md:18]: Please can you keep the important note for kubernetes 1.16 ? That note is still valid for those bumping from kubernetes 1.15.

Something like this :

> Kubernetes 1.16 includes a number of deprecated API removals, and you need to ensure your applications and add ons are updated, or workloads could fail after the upgrade is complete. For more information on the API removals, see the [Kubernetes blog post](https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/). For action you may need to take before upgrading, see the steps in the [EKS documentation](https://docs.aws.amazon.com/eks/latest/userguide/update-cluster.html#1-16-prequisites).

**Comment by barryib** [comment]
> Pushing out a major version every time AWS releases a new version seems a little silly to me. Can't we just make `cluster_version` a required parameter, no default? One last major version bump for this particular issue. Then we'd only need to update the examples which can be a minor version bump.
> 
> @barryib @max-rocket-internet what do you think?

@dpiddockcmp I don't have a strong opinion on that. Maybe @max-rocket-internet or @antonbabenko have one ? But i think that we shouldn’t be afraid of bumping a major release as long as our documentation and module are consistent for a particular EKS version.

> @dpiddockcmp In the end, it's your guys' module. I don't mind the version bump personally, but if it's decided to get rid of it, then 🤷‍♂️

@sc250024 To me it's a community module. It's belong to the community. We are only actual maintainers. That's why @dpiddockcmp is starting a discussion, otherwise he would just drop the default value without asking.
 
Again thank you for your contributions and more broadly to the whole community ❤️❤️❤️

File: README.md (hunk 1 above)
Comment (sc250024) [reply README.md:18]: Will update now.

Comment (sc250024) [reply README.md:18]: done

**Comment by sc250024** [comment]
@barryib I made the doc changes you requested, and pushed. Would you mind re-reviewing please?

**Comment by barryib** [comment]
> @barryib I made the doc changes you requested, and pushed. Would you mind re-reviewing please?

Thanks @sc250024 for your contribution.

Let's wait a bit for @max-rocket-internet or @antonbabenko thought about removing the cluster_version default value.

**Comment by sc250024** [comment]
@barryib Alright.

**Comment by antonbabenko** [comment]
LGTM, too. Thanks, @sc250024

**Comment by abdennour** [comment]
LGTM, too. What is the blocker of the merge here ?

**Comment by abdennour** [comment]
In the meantime, I updated with eksctl : 

```sh
kubectl scale deployments/cluster-autoscaler --replicas=0 -n kube-system

eksctl upgrade nodegroup --name=node-group-name --cluster=cluster-name --kubernetes-version=1.17
```

Good luck!

**Comment by sc250024** [comment]
@max-rocket-internet Anything to add?

**Comment by ldemailly** [comment]
gentle-ping :)

**Comment by max-rocket-internet** [comment]
> Let's wait a bit for @max-rocket-internet or @antonbabenko thought about removing the cluster_version default value.

oooh I could go either way on this one, I don't think it's a big deal to remove the default.....@dpiddockcmp makes good points as always, I think it sounds like a good idea, it would reduce amount of releases and could also perhaps stop a few issues from people asking for the module to "support" the new version (https://github.com/terraform-aws-modules/terraform-aws-eks/issues/775 https://github.com/terraform-aws-modules/terraform-aws-eks/issues/856)

👍

**Comment by ldemailly** [comment]
@barryib lift your request for change/can we get this in? (or is a change needed?)

**Comment by sc250024** [comment]
So this PR has been open for almost a month. I'm fine with making changes, but someone had to pull the trigger. What's the verdict?

**Comment by ldemailly** [comment]
:(

**Comment by arashkaffamanesh** [comment]
@sc250024 does it mean 1.17 is supported with the latest release here?
https://github.com/terraform-aws-modules/terraform-aws-eks/releases/tag/v12.2.0

Or is eksctl still needed as mentioned by @abdennour :

```
eksctl upgrade nodegroup --name=node-group-name --cluster=cluster-name --kubernetes-version=1.17
```

**Comment by filipedeo** [comment]
You'd still need eksctl @arashkaffamanesh, changes were not merged in.

**Comment by abdennour** [comment]
@arashkaffamanesh  .. see answer of @filipedeo

**Comment by cippaciong** [comment]
My 2 cents: I was able to upgrade different EKS clusters to version 1.17 using the latest release of this module (`12.2`). All of them were configured to use managed node groups but I wasn't affected by the issue described by @gdurandvadas even though in some cases I used AWS provider versions that were older than `2.63.0`. The clusters were in different projects and used different configurations and in particular I used these provider versions:
- `2.70.0`
- `2.61.0`
- `2.56.0`

One thing I should mention though, is that when I upgrade my cluster using Terraform, only the control plane is upgraded and I need to upgrade the AMIs of the worker nodes manually, using the AWS console. I've always assumed this was a limitation of managed node groups but looking at @gdurandvadas comments I'm wondering if I'm doing anything wrong in my configuration and I could managed worker nodes AMIs using Terraform.

Just as a reference, here is an example of the configuration I usually use in my EKS clusters.

```hcl
module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  version = "~> 12.2"

  cluster_version = "1.17"
  cluster_name    = var.eks_cluster_name
  subnets         = var.private_subnets

  enable_irsa = true

  tags = {
    Company     = var.company
    Project     = var.project
    Environment = var.environment
    Terraform   = true
  }

  vpc_id = var.vpc_id

  node_groups_defaults = {
    ami_type  = "AL2_x86_64"
    disk_size = 50
  }

  node_groups = {
    base-group = {
      desired_capacity = 4
      max_capacity     = 5
      min_capacity     = 3

      autoscaling_enabled   = true
      protect_from_scale_in = true

      instance_type = "t3a.medium"

      k8s_labels = {
        company     = var.company
        project     = var.project
        environment = var.environment
        terraform   = true
      }
      additional_tags = {
        Company     = var.company
        Project     = var.project
        Environment = var.environment
        Terraform   = true
        # Required by IRSA autoscaler
        "k8s.io/cluster-autoscaler/enabled"                 = true
        "k8s.io/cluster-autoscaler/${var.eks_cluster_name}" = true
      }
    }
  }

  map_roles = var.map_roles
}

```

**Comment by max-rocket-internet** [comment]
Not sure why the PR was closed and branch deleted? But AFAIK, there's nothing stopping anyone using this module to create/upgrade to 1.17.

MNG issue mentioned [here](https://github.com/terraform-aws-modules/terraform-aws-eks/pull/949#issuecomment-659939281) is not related to this module.

Or am I misunderstanding?

**Comment by arashkaffamanesh** [comment]
I can confirm 1.17.9 is working fine with aws provider 2.70.0 and with the cluster-autoscaler and spot-instances, the only issue which I had was getting the autoscaler deployed with helm to work, somehow it didn't like to work and I had to use an older implementation from 1.15. with it, which worked out-of-the-box.

@cippaciong thanks so much for the hint about the provider aws  = "2.70.0"!

```
➜  k get nodes                                                                                                                                                  
NAME                                          STATUS   ROLES    AGE    VERSION
ip-10-2-90-58.eu-central-1.compute.internal   Ready    <none>   172m   v1.17.9-eks-4c6976
```

```
terraform {
  required_version = ">= 0.12.9"

  required_providers {
    # aws        = ">= 2.55.0"
    aws        = "2.70.0"
    local      = ">= 1.4"
    null       = ">= 2.1"
    template   = ">= 2.1"
    random     = ">= 2.1"
    kubernetes = ">= 1.11.1"
  }
}
```

**Comment by ldemailly** [comment]
> Not sure why the PR was closed and branch deleted?
guessing because of non responsiveness of reviewers?

**Comment by sc250024** [comment]
> > Not sure why the PR was closed and branch deleted?
> 
> I am guessing because of non responsiveness of reviewers?

Pretty much.

**Comment by barryib** [comment]
> > > Not sure why the PR was closed and branch deleted?
> > 
> > 
> > I am guessing because of non responsiveness of reviewers?
> 
> Pretty much. And it appears #972 is the preferred solution anyway.

@sc250024 Was in vacation for almost a month. Sorry for the delay.

For the records, we don't need this PR to upgrade to EKS 1.17. Just use the latest version of the module and set your `cluster_version` to `1.17`. In the following version of this module, the cluster version will be required. See https://github.com/terraform-aws-modules/terraform-aws-eks/pull/972

**Pull request closed without merging**